packages/playground/website-cors-proxy/proxy-functions.php

```diff
@@ -0,0 +1,216 @@
+<?php
+
+class CorsProxyException extends Exception
+{
+}
+
+function get_target_url($server_data=null) {
+    if ($server_data === null) {
+        $server_data = $_SERVER;
+    }
+    $requestUri = $server_data['REQUEST_URI'];
+    $targetUrl = $requestUri;
+
+    // Remove the current script name from the beginning of $targetUrl
+    if (strpos($targetUrl, $server_data['SCRIPT_NAME']) === 0) {
+        $targetUrl = substr($targetUrl, strlen($server_data['SCRIPT_NAME']));
+    }
+
+    // Remove the leading slash
+    if ($targetUrl[0] === '/') {
+        $targetUrl = substr($targetUrl, 1);
+    }
+
+    return $targetUrl;
+}
+
+function get_current_script_uri($targetUrl, $request_uri)
+{
+    return substr($request_uri, 0, -strlen($targetUrl));
+}
+
+function url_validate_and_resolve($url, $resolve_function='gethostbynamel') {
+    if (!filter_var($url, FILTER_VALIDATE_URL)) {
+        throw new CorsProxyException("Invalid URL: " . $url);
+    }
+    
+    // Parse the URL to get its components
+    $parsedUrl = parse_url($url);
+
+    // Allow only http and https protocols
+    if (!in_array($parsedUrl['scheme'], ['http', 'https'])) {
+        throw new CorsProxyException("Invalid protocol: " . $parsedUrl['scheme']);
+    }
+
+    // Reject URLs containing username or password before the hostname
+    if (isset($parsedUrl['user']) || isset($parsedUrl['pass'])) {
+        throw new CorsProxyException("URL containing forbidden user or password information");
+    }
+
+    $host = $parsedUrl['host'];
+
+    // Ensure the hostname does not resolve to a private IP
+    $resolved_ips = $resolve_function($host);
+    if ($resolved_ips === false) {
+        throw new CorsProxyException("Hostname could not be resolved");
+    }
+
+    foreach ($resolved_ips as $ip) {
+        if (is_private_ip($ip)) {
+            throw new CorsProxyException("Private IPs are forbidden");
+        }
+    }
+
+    return [
+        'host' => $host,
+        'ip' => $resolved_ips[0]
+    ];
+}
+
+function is_private_ip($ip) {
+    return IpUtils::isPrivateIp($ip);
+}
+
+class IpUtils
+{
+    /**
+     * Checks if the given IP address is a private IP address.
+     *
+     * @param string $ip
+     * @return bool
+     */
+    public static function isPrivateIp($ip)
+    {
+        if (filter_var($ip, FILTER_VALIDATE_IP, FILTER_FLAG_IPV4)) {
+            return self::isPrivateIpv4($ip);
+        } elseif (filter_var($ip, FILTER_VALIDATE_IP, FILTER_FLAG_IPV6)) {
+            return self::isPrivateIpv6($ip);
+        }
+        
+        return false;
+    }
+
+    /**
+     * Checks if the given IPv4 address is private.
+     *
+     * @param string $ip
+     * @return bool
+     */
+    private static function isPrivateIpv4($ip)
+    {
+        // Private IPv4 address ranges
+        $privateRanges = [
```

Comment:
Here is a list of all private IP ranges

IPv4
```
0.0.0.0–0.255.255.255
10.0.0.0–10.255.255.255
100.64.0.0–100.127.255.255
127.0.0.0–127.255.255.255
169.254.0.0–169.254.255.255
172.16.0.0–172.31.255.255
192.0.0.0–192.0.0.255
192.0.2.0–192.0.2.255
192.88.99.0–192.88.99.255
192.168.0.0–192.168.255.255
198.18.0.0–198.19.255.255
198.51.100.0–198.51.100.255
203.0.113.0–203.0.113.255
224.0.0.0–239.255.255.255
233.252.0.0–233.252.0.255
240.0.0.0–255.255.255.254
255.255.255.255
```

IPv6
```
::
::1
::ffff:0.0.0.0 ::ffff:0:0
::ffff:0:0.0.0.0 ::ffff:0:0:0
64:ff9b::0.0.0.0 64:ff9b::0:0
64:ff9b:1::
100::
2001::
2001:20::
2001:db8::
2002::
5f00::
fc00::
fe80::
ff00::
```

Reply:
Thanks! What is the source of that information?

Reply:
I found useful references and more ranges on wikipedia:

https://en.wikipedia.org/wiki/Reserved_IP_addresses